### PR TITLE
rbd-mirror: stop local journal replayer first during shut down

### PR DESCRIPTION
**Pull request by dillaman**
Shutting down the local journal replayer will flush any IO and cancel any
potentially stuck ops (waiting for FinishOp event). This reverts back to the
pre-Octopus behavior prior to the refactoring to support snapshot mirroring.

Fixes: https://tracker.ceph.com/issues/45714
Signed-off-by: Jason Dillaman <dillaman@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman Don't you think it will re-introduce [1]? The reason why the order was changed was that the replay shut down could be initiated when a replay flush was in progress (which also does the shut down).

[1] https://tracker.ceph.com/issues/45409

**Comment by dillaman** [comment]
Forgot about that. We need to keep this order but I suppose that just means we need additional tracking for flushes added.

**Comment by dillaman** [comment]
Tweaked

File: src/tools/rbd_mirror/image_replayer/journal/Replayer.cc [hunk 1]
```diff
@@ -405,14 +405,37 @@ bool Replayer<I>::notify_init_complete(std::unique_lock<ceph::mutex>& locker) {
...
+    m_threads->work_queue, create_context_callback<
+      Replayer<I>, &Replayer<I>::handle_wait_for_flush>(this));
+  m_flush_tracker.wait_for_ops(ctx);
+}
+
+template <typename I>
+void Replayer<I>::handle_wait_for_flush(int r) {
+  dout(10) << "r=" << r << dendl;
+
+  shut_down_local_journal_replay();
```
Comment (trociny) [inline src/tools/rbd_mirror/image_replayer/journal/Replayer.cc:423]: Just thinking. Isn't it theoretically possible that a new replay flush comes in the time window between the flush tracker's `wait_for_ops` completes and the shutdown started?

Comment (dillaman) [reply src/tools/rbd_mirror/image_replayer/journal/Replayer.cc:423]: I think you cannot start a flush if the state is not in `REPLAYING` mode -- and the first thing shut-down does it switch the state.

Comment (trociny) [reply src/tools/rbd_mirror/image_replayer/journal/Replayer.cc:423]: indeed

**Pull request merged**